pkg/gateway/controller.go

```diff
@@ -0,0 +1,203 @@
+	workers        = 5
+)
+
+type Controller struct {
+	gwClient            gatewayclient.Interface
+	gatewayLister       gatewaylisters.GatewayLister
+	gatewayListerSynced cache.InformerSynced
+	gatewayqueue        workqueue.TypedRateLimitingInterface[string]
+
+	httprouteLister       gatewaylisters.HTTPRouteLister
+	httprouteListerSynced cache.InformerSynced
+	httproutequeue        workqueue.TypedRateLimitingInterface[string]
+
+	grpcrouteLister       gatewaylisters.GRPCRouteLister
+	grpcrouteListerSynced cache.InformerSynced
+	grpcroutequeue        workqueue.TypedRateLimitingInterface[string]
+}
+
+func New(
+	gwClient *gatewayclient.Clientset,
+	gatewayInformer gatewayinformers.GatewayInformer,
+	httprouteInformer gatewayinformers.HTTPRouteInformer,
+	grpcrouteInformer gatewayinformers.GRPCRouteInformer,
+) (*Controller, error) {
+	c := &Controller{
+		gwClient:            gwClient,
+		gatewayLister:       gatewayInformer.Lister(),
+		gatewayListerSynced: gatewayInformer.Informer().HasSynced,
+		gatewayqueue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "gateway"},
+		),
+		httprouteLister:       httprouteInformer.Lister(),
+		httprouteListerSynced: httprouteInformer.Informer().HasSynced,
+		httproutequeue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "httproute"},
+		),
+		grpcrouteLister:       grpcrouteInformer.Lister(),
+		grpcrouteListerSynced: grpcrouteInformer.Informer().HasSynced,
+		grpcroutequeue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "grpcroute"},
+		),
+	}
+
+	_, err := gatewayInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			gw := obj.(*gatewayv1.Gateway)
+			if gw.Spec.GatewayClassName != gwClassName {
+				return
+			}
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.gatewayqueue.Add(key)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			gw := newObj.(*gatewayv1.Gateway)
+			if gw.Spec.GatewayClassName != gwClassName {
+				return
+			}
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(newObj)
+			if err == nil {
+				c.gatewayqueue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			gw := obj.(*gatewayv1.Gateway)
+			if gw.Spec.GatewayClassName != gwClassName {
+				return
+			}
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.gatewayqueue.Add(key)
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = httprouteInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```

Comment:
Just like how GatewayClass is used to filter in the Gateway's event handler, HTTPRoutes and GRPCRoutes should also be reconciled only for resources associated with their Gateway (GatewayClass). Otherwise, multiple Gateway API controllers could end up operating on the same resource.

Is it common to look at the [ParentRef](https://gateway-api.sigs.k8s.io/reference/spec/#parentreference) in HTTPRoute or GRPCRoute and then check the GatewayClass of the referenced Gateway?

Reply:
added it now, PTAL